### PR TITLE
[MOD-6771] Distance Calculation using AVX512_BF16

### PR DESCRIPTION
**Pull request by meiravgri**
This pull request introduces an implementation for computing the distance between two vectors utilizing the `AVX512_BF16` instruction set available on Intel CPUs. 
- Note: `AVX512_BF16` only supports bf16 <-> fp32 conversion and dot product operations, not addition or subtraction.

## Key Changes:
- Added AVX512_BF16 Support for **IP** Distance Calculations:

## L2 Distance Calculation (Discarded):
Attempted to implement L2 distance using the formula:
`L2 = sum((v1-v2)^2) = sum(v1^2 - 2v1v2 + v2^2)`
However, due to the lack of support for direct addition and subtraction operations in AVX512_BF16, this method **found to be slower** than the `AVX512BW_VBMI2` flag, and for some cases even slower than the `AVX2` flag.

- For vectors with dimensions **not divisible** by 32, the `AVX512_BF16` implementation performed even slower than `AVX2`.
- For vectors with no residual, the `AVX512_BF16` showed a slight performance improvement over `AVX2`.
- Despite the minor superiority for non-residual dimensions, the code complexity and limited performance gains **did not justify** maintaining the `AVX512_BF16` optimization for the L2 distance function.
image

image

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Review by GuyAv46 — DISMISSED** [review]

File: src/VecSim/spaces/IP/IP_AVX512_BF16_VL_BF16.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        pVect2 += residual;
+        sum = _mm512_dpbf16_ps(sum, (__m512bh)v1, (__m512bh)v2);
+    }
+
+    do {
+        InnerProductStep(pVect1, pVect2, sum);
+    } while (pVect1 < pEnd1);
+
+    return 1.0f - _mm512_reduce_add_ps(sum);
+}
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512_BF16_VL_BF16.h:45]: Cleaaaan

File: src/VecSim/spaces/IP/IP_AVX512_BF16_VL_BF16.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "VecSim/spaces/space_includes.h"
+#include "VecSim/types/bfloat16.h"
+
+using bfloat16 = vecsim_types::bfloat16;
+
+static inline void InnerProductStep(bfloat16 *&pVect1, bfloat16 *&pVect2, __m512 &sum) {
+    __m512i vec1 = _mm512_loadu_si512((__m512i *)pVect1);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512_BF16_VL_BF16.h:13]: nope 🤷

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/IP/IP_AVX512_BF16_VL_BF16.h:13]: only for 128/256 bits

**Review by alonre24 — DISMISSED** [review]
Niceee :)

**Review by GuyAv46 — APPROVED** [review]

**Pull request merged**